Add FluxPAGPipeline with support for PAG

tongyu0924 opened the pull request:
# What does this PR do?

This PR adds support for Perturbed Attention Guidance (PAG) to the FluxPipeline

Fixes #11488 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @yiyixuxu
- Pipelines and pipeline callbacks: @yiyixuxu and @asomoza
- Training examples: @sayakpaul
- Docs: @stevhliu and @sayakpaul
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @sayakpaul @yiyixuxu @DN6

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc
- PEFT: @sayakpaul @BenjaminBossan

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

DN6 commented:
@tongyu0924 Is this ready for review? If so, could you move the pipeline under `src/diffusers/pipelines/pag` please and make the changes needed to import the pipeline?

tongyu0924 commented:
> @tongyu0924 Is this ready for review? If so, could you move the pipeline under `src/diffusers/pipelines/pag` please and make the changes needed to import the pipeline?

Done! The pipeline is now under src/diffusers/pipelines/pag. It's ready for review

DN6 commented:
Thank you @tongyu0924 👍🏽!  Could we add the pipeline to the necessary init files

PAG Module:
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/__init__.py
 
Pipelines Module:
https://github.com/huggingface/diffusers/blob/049082e013fb71d78f3abf487916f3de2b674908/src/diffusers/pipelines/__init__.py#L175

Diffusers main init
https://github.com/huggingface/diffusers/blob/049082e013fb71d78f3abf487916f3de2b674908/src/diffusers/__init__.py#L341

And then could you please add a fast test for the pipeline, similar to how it has been done here
https://github.com/huggingface/diffusers/blob/049082e013fb71d78f3abf487916f3de2b674908/tests/pipelines/pag/test_pag_sd3.py#L26

tongyu0924 commented:
> Thank you @tongyu0924 👍🏽! Could we add the pipeline to the necessary init files
> 
> PAG Module: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/__init__.py
> 
> Pipelines Module:
> 
> https://github.com/huggingface/diffusers/blob/049082e013fb71d78f3abf487916f3de2b674908/src/diffusers/pipelines/__init__.py#L175
> 
> Diffusers main init
> 
> https://github.com/huggingface/diffusers/blob/049082e013fb71d78f3abf487916f3de2b674908/src/diffusers/__init__.py#L341
> 
> And then could you please add a fast test for the pipeline, similar to how it has been done here
> 
> https://github.com/huggingface/diffusers/blob/049082e013fb71d78f3abf487916f3de2b674908/tests/pipelines/pag/test_pag_sd3.py#L26

I've added the pipeline to the necessary `__init__.py` files and added a fast test following the structure in `test_pag_sd3.py`.

nitinmukesh commented:
Please update the following if possible

diffusers/src/diffusers/pipelines/pag/pipeline_pag_flux.py

EXAMPLE_DOC_STRING = """
    Examples:
        ```py
        >>> import torch
        >>> from diffusers import FluxPipeline

        >>> pipe = FluxPipeline.from_pretrained("black-forest-labs/FLUX.1-schnell", torch_dtype=torch.bfloat16)
        >>> pipe.to("cuda")
        >>> prompt = "A cat holding a sign that says hello world"
        >>> # Depending on the variant being used, the pipeline call will slightly vary.
        >>> # Refer to the pipeline documentation for more details.
        >>> image = pipe(prompt, num_inference_steps=4, guidance_scale=0.0).images[0]
        >>> image.save("flux.png")
        ```
"""

semioz commented:
hi, just wanted to check in, are there any updates on when this might be merged?
